### PR TITLE
[Azure Service Bus] Batch.TryAdd should respect service limits of 4500 items max

### PR DESCRIPTION
**Pull request by danielmarbach**
Adhere to the service limits

### CODE REVIEW

**Comment by danielmarbach** [comment]
@JoshLove-msft @jongio @ramya-rao-a while doing my talk update I stumbled over a service error when trying to send multiple thousand messages in a batch. My research let me down to discover this comment https://github.com/Azure/azure-sdk-for-net/pull/19091/files#diff-0ab95e5bde2b334bbabc6679c5ed6777bd349b06d07eeec4a040b6d681e82ee6R42

So I figured I send in a draft PR to address this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs [hunk 1]
```diff
@@ -118,15 +118,15 @@ public override bool TryAddMessage(ServiceBusMessage message)
                         ? OverheadBytesSmallMessage
                         : OverheadBytesLargeMessage);
 
-                if (size > MaxSizeInBytes)
+                if (size > MaxSizeInBytes || BatchMessages.Count >= Options.MaxSizeInNumberOfItems.Value)
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:121]: Wasn't entirely sure if you would prefer to copy the value to a field as done for

```
MaxSizeInBytes = options.MaxSizeInBytes.Value;
```

given that we also store the options locally I went with accessing the options directly

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs [hunk 2]
```diff
@@ -118,15 +118,15 @@ public override bool TryAddMessage(ServiceBusMessage message)
...
                 {
                     return false;
                 }
 
                 _sizeBytes = size;
                 BatchMessages.Add(message);
 
-            return true;
+                return true;
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:129]: Can be reverted if not wanted. It just looked awkward and I felt it is not a big sin to change this. Also happy to pull it into a dedicated commit if prefered

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/CreateMessageBatchOptions.cs [hunk 3]
```diff
@@ -40,6 +43,30 @@ public class CreateMessageBatchOptions
...
+        ///
+        internal int? MaxSizeInNumberOfItems
+        {
+            get => _maxSizeInNumberOfItems;
+
+            set
+            {
+                if (value.HasValue)
+                {
+                    Argument.AssertInRange(value.Value, 1, 4500, nameof(MaxSizeInNumberOfItems));
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/CreateMessageBatchOptions.cs:63]: I doubt having a batch size of one ever makes sense yet I had to get started with something.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/CreateMessageBatchOptions.cs [hunk 4]
```diff
@@ -40,6 +43,30 @@ public class CreateMessageBatchOptions
...
+        /// <summary>
+        ///   The maximum size to allow for a single batch of messages, in number of items.
+        /// </summary>
+        ///
+        /// <value>
+        ///   The desired limit, in number of items, for the size of the associated service bus message batch.  If <c>null</c>,
+        ///   the maximum size allowed by the active transport will be used.
+        /// </value>
+        ///
+        internal int? MaxSizeInNumberOfItems
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/CreateMessageBatchOptions.cs:55]: At first, I had this public then I figured I'll start with making it internal first. Yet I guess this could be a nice add if wanted to the public API.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/CreateMessageBatchOptions.cs [hunk 5]
```diff
@@ -16,6 +16,9 @@ public class CreateMessageBatchOptions
         /// <summary>The requested maximum size to allow for the batch, in bytes.</summary>
         private long? _maxSizeInBytes;
 
+        /// <summary>The requested maximum size to allow for the batch, in number of items.</summary>
+        private int? _maxSizeInNumberOfItems = 4500;
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/CreateMessageBatchOptions.cs:20]: Maybe this constant value belongs somewhere else. Please advisese

**Comment by danielmarbach** [comment]
The naive code I used was

```
            var messagesToSend = new Queue<ServiceBusMessage>();
            for (var i = 0; i < 6500; i++)
            {
                var message = new ServiceBusMessage($"Deep Dive{i}");
                messagesToSend.Enqueue(message);
            }
```

which caused

```
Sending 5849 messages in a batch 1.
Unhandled exception. Azure.Messaging.ServiceBus.ServiceBusException: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 Reference:7a26ac50-cc53-4fb0-9f
2d-c84f90fc6057, TrackingId:bc624f8e00000025006f83d2603e32c7_G10_B9, SystemTracker:nservicebustests:Queue:queue, Timestamp:2021-03-02T12:43:02 (GeneralError)
   at Azure.Messaging.ServiceBus.Amqp.AmqpSender.SendBatchInternalAsync(Func`1 messageFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.Amqp.AmqpSender.SendBatchInternalAsync(Func`1 messageFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.Amqp.AmqpSender.<>c__DisplayClass19_0.<<SendBatchAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation(Func`2 operation, TransportConnectionScope scope, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation(Func`2 operation, TransportConnectionScope scope, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.Amqp.AmqpSender.SendBatchAsync(ServiceBusMessageBatch messageBatch, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ServiceBusSender.SendMessagesAsync(ServiceBusMessageBatch messageBatch, CancellationToken cancellationToken)
   at Batching.Program.Main(String[] args) in C:\p\AzureServiceBus.DeepDive\Batching\Program.cs:line 106
   at Batching.Program.Main(String[] args) in C:\p\AzureServiceBus.DeepDive\Batching\Program.cs:line 87
   at Batching.Program.Main(String[] args) in C:\p\AzureServiceBus.DeepDive\Batching\Program.cs:line 87
   at Batching.Program.Main(String[] args) in C:\p\AzureServiceBus.DeepDive\Batching\Program.cs:line 87
   at Batching.Program.<Main>(String[] args)
```

**Comment by jsquire** [comment]
Hi @danielmarbach.   Thank you very much for your contribution and the efforts that you continue to put into improving the SDK.  Generally, we try to eagerly validate items that have meaning only to the SDK and based on information that the service provides to us  and avoid making assumptions about service logic that may change.

In this case, the service doesn't communicate an object limit, so we'd be mirroring logic that exists on the service and that may change in the future.   I believe that we've talked over this validation within the SDK team a couple of times and reached the conclusion that this was enough of an edge case that we didn't feel justified the risk having the client be out-of-sync with the service and enforcing the wrong value.    _(I'll ask the folks called out below to make sure my recollection here is correct.)_

I don't know that we discussed asking the service team to consider adding this limit as a property on the sending link when opened, similar to how they provide the maximum allowable size for a message.   That would allow us to validate while leaving the service in control over the value.   Thoughts?

//cc: @JoshLove-msft, @ramya-rao-a

**Comment by danielmarbach** [comment]
In general, I agree with the reasoning and think it is a good design approach. Where I see problems though with the current API from a user perspective is that I assume that `TryAdd` does the right thing for me but then I run into a weird `GeneralError` that doesn't really tell me as a user what's going on. 

On top of that, it might be likely that you don't find this out while testing with smaller batch sizes but if you determine the number of messages to be sent based on a dynamic set that might be larger you suddenly run into surprises in production. And when you do, again, the error doesn't really tell you what is going on.

**Comment by JoshLove-msft** [comment]
> In general, I agree with the reasoning and think it is a good design approach. Where I see problems though with the current API from a user perspective is that I assume that `TryAdd` does the right thing for me but then I run into a weird `GeneralError` that doesn't really tell me as a user what's going on.
> 
> On top of that, it might be likely that you don't find this out while testing with smaller batch sizes but if you determine the number of messages to be sent based on a dynamic set that might be larger you suddenly run into surprises in production. And when you do, again, the error doesn't really tell you what is going on.

I agree that the current behavior will be for confusing for users. I think the main reason we didn't add the validation is that this 4500 limit is actually different between standard and premium namespaces. /cc @HarshaNalluru

**Comment by SeanFeldman** [comment]
Do not want to derail the conversation, just a remark that could be handled separately if the team wants about 

> I agree that the current behavior will be for confusing for users. I think the main reason we didn't add the validation is that this 4500 limit is actually different between standard and premium namespaces

These discrepancies between Standard and Premium are only going to increase. The maximum number of items (this discussion), partitioning, maximum message size, etc.

> We can check with the service team to see if there is a feature request to include the max number of messages as a property on the link.

Potentially, asking the question about further features that will be different between the tiers and impacting the client SDK.

**Comment by ramya-rao-a** [comment]
> These discrepancies between Standard and Premium are only going to increase. The maximum number of items (this discussion), partitioning, maximum message size, etc.

Fortunately the maximum message size is something that the client is privy to as we can read that information from the AMQP link properties once it has been established. The ideal way forward would be to get the max message count information in the same way. We will circle back here once we get an update from the service team on any plans to do so or thoughts on alternatives.

**Comment by danielmarbach** [comment]
Well I guess the ball is in the hands of the team :)

**Comment by SeanFeldman** [comment]
Paging @JoshLove-msft and @jsquire directly as they are marked as reviewers.

**Comment by JoshLove-msft** [comment]
Sorry about that bot reply - we are trying to see if the service can expose the max message count to the client. We will update once we have more information on this.

**Comment by jsquire** [comment]
I've lost track around conversations for this ask and have opened #21451 to give us a stable place to track.
